presto-main/src/main/java/com/facebook/presto/execution/scheduler/FixedSubsetNodeSetSupplier.java

```diff
@@ -0,0 +1,296 @@
+import com.facebook.presto.metadata.InternalNodeManager;
+import com.facebook.presto.spi.ConnectorId;
+import com.facebook.presto.spi.HostAddress;
+import com.facebook.presto.spi.Node;
+import com.facebook.presto.spi.QueryId;
+import com.google.common.base.Supplier;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSetMultimap;
+
+import javax.inject.Inject;
+import javax.validation.constraints.NotNull;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.PriorityQueue;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.execution.scheduler.NodeSelectionHashStrategy.CONSISTENT_HASHING;
+import static com.facebook.presto.metadata.InternalNode.NodeStatus.ALIVE;
+import static com.facebook.presto.spi.NodeState.ACTIVE;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static java.lang.Math.ceil;
+import static java.lang.String.format;
+
+/**
+ * This WIP class is responsible for assigning a subset of the cluster nodes
+ * to a requesting query.
+ * The size of the subset is currently static, but it can by defined by configuration
+ * or can be driven by fairness algorithm depending on query metadata.
+ */
+public class FixedSubsetNodeSetSupplier
+        implements NodeSetSupplier, Closeable
+{
+    private static final Logger log = Logger.get(FixedSubsetNodeSetSupplier.class);
+    private final PriorityQueue<NodeSetAcquireRequest> pendingRequests;
+    private final InternalNodeManager nodeManager;
+    private final NetworkLocationCache networkLocationCache;
+    // TODO: Use Set<InternalNode>
+    private final Map<QueryId, List<InternalNode>> perQueryNodeAssignments;
+
+    private final ScheduledFuture<?> scheduledFuture;
+
+    @Inject
+    public FixedSubsetNodeSetSupplier(
+            InternalNodeManager nodeManager,
+            NetworkTopology networkTopology,
+            @ForNodeScheduler ScheduledExecutorService scheduledExecutorService)
+    {
+        this.nodeManager = nodeManager;
+        this.networkLocationCache = new NetworkLocationCache(networkTopology);
+        this.perQueryNodeAssignments = new HashMap<>();
+        this.pendingRequests = new PriorityQueue<>();
+        // start loop to check pending node requests every second
+        scheduledFuture = scheduledExecutorService.scheduleAtFixedRate(this::fulfillPendingRequests, 0, 100, TimeUnit.MILLISECONDS);
```

Comment:
nit: To reduce latency we may want to trigger `fulfillPendingRequests` each time `acquire` / `release` is called.

Reply:
Makes sense. Thats what I originally had in mind but took a shortcut for POC. But you are right we probably need to fix this for POC as well. Will update now